### PR TITLE
[Job Launcher] Added refund payment logic

### PR DESCRIPTION
**Pull request by eugenvoronov**
## Description
Job Launcher needs to refund the remaining balance in the contract we job requester asks to cancel an escrow.

## Summary of changes
1. Canceled the escrow.
2. Created new payment type called REFUND.
3. Created new job type called TO_REFUND.
4. From logs got the amount received back from the escrow contract.
5. Created a REFUND payment for the Job Requester.
6. Updated job cancelation logic.
7. Added payment refund logic.

## How test the changes
1. `POST /job/fortune`
2. `GET /job/cron/launch`
3. `PATCH /job/cancel/:jobId`
4. `GET /job/cron/cancel`
5. `GET /job/cron/refund`

`yarn test`

## Related issues
[[Job Launcher] Refund in cancellation#843](https://github.com/humanprotocol/human-protocol/issues/843)

### CODE REVIEW

**Comment by eugenvoronov** [comment]
- Why have you added and endpoint for refund? I mean, you can add a new refund entry for the job when you call the cancel endpoint instead of having the user to make another request.
- The main one it because of blockchain, if you try to get a logs after cancel escrow method you wont be able to find any logs about release, and we wont be able to calculate right refund amount without it.

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts [hunk 1]
```diff
@@ -530,7 +540,140 @@ export class JobService {
...
+      console.log(123, logs)
+      let refundAmount = new Decimal(0);
+
+      logs.forEach(log => {
+          const parsedLog = tokenContract.interface.parseLog(log);
+          const from = parsedLog.args[0];
+          const to = parsedLog.args[1];
+          const amount = parsedLog.args[2];
+
+          console.log(from, escrowAddress, to, signer.address)
```
Comment (flopez7) [inline packages/apps/job-launcher/server/src/modules/job/job.service.ts:649]: remove logs

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts [hunk 2]
```diff
@@ -530,7 +540,140 @@ export class JobService {
     }
   }
 
+  public async refundCronJob() {
+    // TODO: Add retry policy and process failure requests https://github.com/humanprotocol/human-protocol/issues/334
+    const fiveMinutesAgo = new Date(Date.now() - 5 * 60 * 1000);
```
Comment (leric7) [inline packages/apps/job-launcher/server/src/modules/job/job.service.ts:545]: Should be better define 5 Minutes as a constant.

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts [hunk 3]
```diff
@@ -530,7 +540,140 @@ export class JobService {
...
+      const currentBlockNumber = await signer.provider.getBlockNumber();
+      const filter = {
+          address: tokenAddress,
+          topics: [ethers.utils.id('Transfer(address,address,uint256)')],
+          fromBlock: currentBlockNumber - 200,
+          toBlock: 'latest',
+      };
+
+      const logs = await signer.provider.getLogs(filter);
+      console.log(123, logs)
```
Comment (leric7) [inline packages/apps/job-launcher/server/src/modules/job/job.service.ts:640]: Remove console.log

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts [hunk 4]
```diff
@@ -530,7 +540,140 @@ export class JobService {
...
+
+  public async calculateRefundAmount(chainId: ChainId, tokenAddress: string, escrowAddress: string) {
+      const signer = this.web3Service.getSigner(chainId);
+      const tokenContract: HMToken = HMToken__factory.connect(tokenAddress, signer);
+
+      const currentBlockNumber = await signer.provider.getBlockNumber();
+      const filter = {
+          address: tokenAddress,
+          topics: [ethers.utils.id('Transfer(address,address,uint256)')],
+          fromBlock: currentBlockNumber - 200,
```
Comment (leric7) [inline packages/apps/job-launcher/server/src/modules/job/job.service.ts:635]: Better define `fromBlock` threshold of 200 as a constant, and reuse it from `cancelCronJobOld` function. It's now 100 over there, and looks like it's inconsistent.

**Review by leric7 — CHANGES_REQUESTED** [review]
Overall looks good. Just couple NITs.

File: packages/apps/job-launcher/server/src/database/migrations/1691485394906-InitialMigration.ts
Comment (portuu3) [inline packages/apps/job-launcher/server/src/database/migrations/1691485394906-InitialMigration.ts:1]: do not edit initial migrations file, add a new migration

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts [hunk 5]
```diff
@@ -535,90 +542,132 @@ export class JobService {
...
-  public async cancelCronJob() {
+  public async refundCronJob() {
     // TODO: Add retry policy and process failure requests https://github.com/humanprotocol/human-protocol/issues/334
     const jobEntity = await this.jobRepository.findOne(
       {
-        status: JobStatus.TO_CANCEL,
+        status: JobStatus.TO_REFUND,
         retriesCount: LessThanOrEqual(JOB_RETRIES_COUNT_THRESHOLD),
         waitUntil: LessThanOrEqual(new Date()),
+        updatedAt: LessThanOrEqual(_5_MINS)
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/modules/job/job.service.ts:552]: why you add this condition?

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts [hunk 6]
```diff
@@ -535,90 +542,132 @@ export class JobService {
     }
   }
 
-  public async cancelCronJob() {
+  public async refundCronJob() {
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/modules/job/job.service.ts:545]: why we split cancel and refund logic? We need to refund right after cancel when the funds have been received in the job launcher account

**Review by portuu3 — CHANGES_REQUESTED** [review]

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts (hunk 5 above)
Comment (eugenvoronov) [reply packages/apps/job-launcher/server/src/modules/job/job.service.ts:552]: @portuu3

Comment (eugenvoronov) [reply packages/apps/job-launcher/server/src/modules/job/job.service.ts:552]: This period ensures that the blockchain has successfully processed the cancellation request. It verifies that the escrow was canceled and confirms the funds were transferred once the job was initiated.

File: packages/apps/job-launcher/server/src/modules/job/job.service.ts (hunk 6 above)
Comment (eugenvoronov) [reply packages/apps/job-launcher/server/src/modules/job/job.service.ts:545]: https://github.com/humanprotocol/human-protocol/pull/869#issuecomment-1713799233

Comment (eugenvoronov) [reply packages/apps/job-launcher/server/src/modules/job/job.service.ts:545]: If we want to return the amount that will be released from the escrow address, which we can receive in the escrow logs, then the logs will not yet be available.

**Comment by eugenvoronov** [comment]
PR was migrate https://github.com/humanprotocol/human-protocol/pull/928

**Pull request closed without merging**